fdbserver/storageserver.actor.cpp

```diff
@@ -5951,6 +5963,247 @@ ACTOR Future<Void> auditStorageShardReplicaQ(StorageServer* data, AuditStorageRe
+	}
+	return;
+}
+
+// The SS actor handling bulk dump task sent from DD.
+// The SS partitions the task range into batches and make progress on each batch one by one.
+// Each batch is a subrange of the task range sent from DD.
+// When SS completes one batch, SS persists the metadata indicating this batch range completed.
+// If the SS fails on dumping a batch data, the SS will send an error to DD and the leftover files
+// is cleaned up when this actor returns.
+// In the case of SS crashes, the leftover files will be cleared at the init step when the SS restores.
+// If the SS uploads any file with succeed but the blob store is actually stored, this inconsistency will
+// be captured by DD and DD will retry to dump the problematic range with a new task.
+// DD will retry later if it receives any error from SS.
+// Upload the data for the range with the following path organization:
+//  <rootRemote>/<JobId>/<TaskId>/<batchNum>/<dumpVersion>-manifest.sst
+//	<rootRemote>/<JobId>/<TaskId>/<batchNum>/<dumpVersion>-data.sst
+//	<rootRemote>/<JobId>/<TaskId>/<batchNum>/<dumpVersion>-sample.sst
+// where rootRemote = req.bulkDumpState.remoteRoot, jobId = req.bulkDumpState.jobId, taskId = req.bulkDumpState.taskId,
+// batchNum and dumpVersion are dynamically generated.
+// Each task must have one manifest file.
+// If the task's range is empty, data file and sample file do not exist
+// If the task's data size is too small, the sample file may omitted
+ACTOR Future<Void> bulkDumpQ(StorageServer* data, BulkDumpRequest req) {
+	wait(data->serveBulkDumpParallelismLock.take(TaskPriority::DefaultYield));
+	state FlowLock::Releaser holder(data->serveBulkDumpParallelismLock); // A SS can handle one bulkDump task at a time
+	state Key rangeBegin = req.bulkDumpState.getRange().begin;
+	state Key rangeEnd = req.bulkDumpState.getRange().end;
+	state int64_t readBytes = 0;
+	state int retryCount = 0;
+	state int batchNum = 0;
+	state Version versionToDump;
+	state RangeDumpData rangeDumpData;
+	state std::string rootFolderLocal = getBulkDumpLocalRoot(data);
+	// Use jobId and taskId as the folder to store the data of the task range
+	ASSERT(req.bulkDumpState.getTaskId().present());
+	state std::string taskFolder =
+	    generateBulkDumpTaskFolder(req.bulkDumpState.getJobId(), req.bulkDumpState.getTaskId().get());
+	state BulkDumpFileSet destinationFileSets;
+
+	loop {
+		try {
+			if (batchNum == 0) { // clean up working folder initially
+				platform::eraseDirectoryRecursive(abspath(joinPath(rootFolderLocal, taskFolder)));
+			}
+
+			// Dump data of rangeToDump in a relativeFolder
+			state KeyRange rangeToDump = Standalone(KeyRangeRef(rangeBegin, rangeEnd));
+
+			// relativeFolder = <JobId>/<TaskId>/<batchNum>
+			// relativeFolder remains consistent between local path and remote path
+			state std::string relativeFolder = joinPath(taskFolder, std::to_string(batchNum));
+
+			// Get version to dump
+			wait(store(versionToDump, getReadVersion(data)));
```

Comment:
is this necessary? how about just use `data->version`?

Reply:
Yes. In the future, we are going to read from multiple SSes and compare checksums. So, the GRV may be better.